lib/app_jwt/app_jwt.c

```diff
@@ -0,0 +1,767 @@
+	char pubkey_der_str[ECDSA_PUBLIC_KEY_DER_SZ * 2] = {0};
+
+	size_t pubkey_der_str_len = 0;
+
+	err = bytes_to_str(pubkey_der, pubkey_der_len, pubkey_der_str, sizeof(pubkey_der_str),
+			   &pubkey_der_str_len);
+	if (err == 0) {
+		/* Print DER formatted pubkey to traces */
+		LOG_INF("pubkey_der (%d) =  %s", pubkey_der_len, pubkey_der_str);
+	}
+#endif /* CONFIG_APP_JWT_PRINT_EXPORTED_PUBKEY_DER */
+
+	/* Compute the SHA256 hash of public key DER format */
+	status = psa_hash_compute(PSA_ALG_SHA_256, pubkey_der, sizeof(pubkey_der), key_hash_out,
+				  key_hash_buffer_size, key_hash_Length);
+	if (status != PSA_SUCCESS) {
+		LOG_ERR("psa_hash_compute failed! (Error: %d)", status);
+		return -1;
+	}
+
+	return status;
+}
+
+static int sign_message(const uint32_t user_key_id, const uint8_t *input, size_t input_length,
+			uint8_t *signature, size_t signature_size, size_t *signature_length)
+{
+	psa_status_t status;
+
+	/* Sign the hash */
+	status = psa_sign_message(user_key_id, PSA_ALG_ECDSA(PSA_ALG_SHA_256), input, input_length,
+				  signature, signature_size, signature_length);
+	if (status != PSA_SUCCESS) {
+		LOG_ERR("psa_sign_hash failed! (Error: %d)", status);
+		return -1;
+	}
+
+	return 0;
+}
+
+#if defined(CONFIG_APP_JWT_VERIFY_SIGNATURE)
+
+static int verify_message_signature(const uint32_t user_key_id, const char *const message,
+				    size_t message_size, uint8_t *signature, size_t signature_size)
+{
+	psa_status_t status;
+
+	/* Verify the signature of the message */
+	status = psa_verify_message(user_key_id, PSA_ALG_ECDSA(PSA_ALG_SHA_256), message,
+				    message_size, signature, signature_size);
+	if (status != PSA_SUCCESS) {
+		LOG_INF("signature verification failed! (Error: %d)", status);
+		return -1;
+	}
+
+	return 0;
+}
+#endif /* CONFIG_APP_JWT_VERIFY_SIGNATURE */
+
+static int get_random_bytes(uint8_t *output_buffer, const size_t output_length)
+{
+	int err = -ENODATA;
+
+	if ((NULL == output_buffer) || (output_length < 4)) {
+		/* bad parameter */
+		return -EINVAL;
+	}
+
+	psa_status_t psa_status = psa_generate_random(output_buffer, output_length);
+
+	if (PSA_SUCCESS != psa_status) {
+		LOG_ERR("psa_generate_random failed! = %d", psa_status);
+		err = -ENOMEM;
+	} else {
+		err = 0;
+	}
+
+	return err;
+}
+
+static int crypto_finish(void)
+{
+	psa_status_t status;
+
+	/* Purge the key from memory */
+	status = psa_purge_key(IAK_APPLICATION_GEN1);
+	if (status != PSA_SUCCESS) {
+		LOG_ERR("psa_purge_key failed! (Error: %d)", status);
+		return -ENOMEM;
+	}
+
+	return 0;
+}
+
+static char *jwt_header_create(const uint32_t alg, const uint32_t keyid)
+{
+	int err = 0;
+
+	uint8_t pub_key_hash[ECDSA_SHA_256_HASH_SZ];
+
+	char *hdr_str = NULL;
+
+	cJSON *jwt_hdr = cJSON_CreateObject();
+
+	if (!jwt_hdr) {
+		LOG_ERR("cJSON_CreateObject failed!");
+		return NULL;
+	}
+
+	/* Type: format: always "JWT" */
+	if (cJSON_AddStringToObjectCS(jwt_hdr, "typ", "JWT") == NULL) {
+		return NULL;
```

Comment:
This and all following `return NULL;` calls leak memory, because `jwt_hdr` is not freed. Maybe it would be easiest to add a label in the end of the function before the `cJSON_Delete()` call (for example `exit` or `clean_exit`) and when an error happens, `goto` there.